metadata-io/src/main/java/com/linkedin/metadata/entity/DeleteEntityService.java

```diff
@@ -0,0 +1,254 @@
+import javax.annotation.Nonnull;
+import lombok.RequiredArgsConstructor;
+import lombok.extern.slf4j.Slf4j;
+import org.apache.commons.collections.CollectionUtils;
+
+import static com.linkedin.metadata.search.utils.QueryUtils.*;
+
+
+@Slf4j
+@RequiredArgsConstructor
+public class DeleteEntityService {
+
+  private final EntityService _entityService;
+  private final GraphService _graphService;
+
+  private static final Integer ELASTIC_BATCH_DELETE_SLEEP_SEC = 5;
+
+  /**
+   * Public endpoint that deletes references to a given urn across DataHub's metadata graph. This is the entrypoint for
+   * addressing dangling pointers whenever a user deletes some entity.
+   *
+   * @param urn     The urn for which to delete references in DataHub's metadata graph.
+   * @param dryRun  Specifies if the delete logic should be executed to conclusion or if the caller simply wants a
+   *                preview of the response.
+   * @return  A {@link DeleteReferencesResponse} instance detailing the response of deleting references to the provided
+   * urn.
+   */
+  public DeleteReferencesResponse deleteReferencesTo(final Urn urn, final boolean dryRun) {
+    final DeleteReferencesResponse result = new DeleteReferencesResponse();
+    RelatedEntitiesResult relatedEntities =
+        _graphService.findRelatedEntities(null, newFilter("urn", urn.toString()), null,
+            EMPTY_FILTER,
+            ImmutableList.of(),
+            newRelationshipFilter(EMPTY_FILTER, RelationshipDirection.INCOMING), 0, 10000);
+
+    final List<RelatedAspect> relatedAspects = relatedEntities.getEntities().stream().map(relatedEntity -> {
+          final Urn relatedUrn = UrnUtils.getUrn(relatedEntity.getUrn());
+          final String relationType = relatedEntity.getRelationshipType();
+          final String relatedEntityName = relatedUrn.getEntityType();
+          final EntitySpec relatedEntitySpec = _entityService.getEntityRegistry().getEntitySpec(relatedEntityName);
+
+          return getAspectReferringTo(urn, relatedUrn, relationType, relatedEntitySpec).map(pair -> {
+            final RelatedAspect relatedAspect = new RelatedAspect();
+            relatedAspect.setEntity(relatedUrn);
+            relatedAspect.setRelationship(relationType);
+            relatedAspect.setAspect(pair.getSecond().getName());
+            return relatedAspect;
+          });
+        }).filter(java.util.Optional::isPresent)
+        .map(java.util.Optional::get)
+        .limit(10).collect(Collectors.toList());
+
+    result.setRelatedAspects(new RelatedAspectArray(relatedAspects));
+    result.setTotal(relatedEntities.getTotal());
+
+    if (dryRun) {
+      return result;
+    }
+
+    int processedEntities = 0;
+
+    // Delete first 10k
+    relatedEntities.getEntities().forEach(entity -> deleteReference(urn, entity));
+
+    processedEntities += relatedEntities.getEntities().size();
+
+    // Delete until less than 10k are left
+    while (relatedEntities.getCount() > 10000) {
```

Comment:
Bug alert: I think this should be relatedEntities.getTotal()?

Reply:
It's more like a do-while loop but he didn't want to add sleep in the beginning. 
So the intial one tries to do the first query, if that returns less than 10K we are good. If it returns 10K, then we need to sleep for the 10K deletes to have gone through then try again until it returns less than 10K

Reply:
Is this logic acceptable jjoyce?

Reply:
cannot we simply do this in a normal for loop? I don't like how we duplicate the logic inside and outside of the loop. Please think about making it simpler.